src/cljs/athens/util.cljs

```diff
@@ -92,3 +95,71 @@
       (string/replace x #"PM" "pm"))))
 
 
+;; -- Linked & Unlinked References ----------
+
+(defn get-ref-ids
+  [pattern]
+  @(q '[:find [?e ...]
+        :in $ ?regex
+        :where
+        [?e :block/string ?s]
+        [(re-find ?regex ?s)]]
+      db/dsdb
+      pattern))
+
+
+(defn merge-parents-and-block
+  [ref-ids]
+  (let [parents (reduce-kv (fn [m _ v] (assoc m v (db/get-parents-recursively v)))
+                           {}
+                           ref-ids)
+        blocks (map (fn [id] (db/get-block-document id)) ref-ids)]
+    (mapv
+      (fn [block]
+        (merge block {:block/parents (get parents (:db/id block))}))
+      blocks)))
+
+
+(defn group-by-parent
+  [blocks]
+  (group-by (fn [x]
+              (-> x
+                  :block/parents
+                  first
+                  :node/title))
+            blocks))
+
+
+(defn get-data
+  [pattern]
+  (-> pattern get-ref-ids merge-parents-and-block group-by-parent seq))
+
+
+(defn get-data-by-block
+  [pattern]
+  (-> pattern get-ref-ids merge-parents-and-block seq))
+
+
+(defn get-linked-references
+  [title]
+  (-> title patterns/linked get-data))
+
+
+(defn get-linked-references-by-block
+  [title]
+  (-> title patterns/linked get-data-by-block))
+
+
+(defn get-unlinked-references
+  [title]
+  (-> title patterns/unlinked get-data))
+
+
+(defn count-linked-references-excl-uid
```

Comment:
If I reasoned correctly, because dispatch is async, simply decreasing count by one could result in a race condition that might or might not return one less reference than expected.